### PR TITLE
Refactor TLS configuration

### PR DESCRIPTION
**Pull request by cw9**
Refactor the logic around auth to keep the tls related logic isolated from etcd cluster logic

@xichen2020 @prateek @jeromefroe

### CODE REVIEW

File: client/etcd/config.go [hunk 1]
```diff
@@ -26,16 +26,28 @@ import (
...
-	CA        string   `yaml:"ca",omitempty`
+	Zone      string     `yaml:"zone"`
+	Endpoints []string   `yaml:"endpoints"`
+	Auth      AuthConfig `yaml:"auth"`
 }
 
-// Configuration is for config service client
+// AuthConfig is the config for authentication.
+type AuthConfig struct {
+	Cert string `yaml:"cert"`
```
Comment (prateek) [inline client/etcd/config.go]: A few naming nits:

- rename to TLSConfig
- change var names to reflect what they are (path) and the type of file (crt/key).

Something like this:

```
type TLSConfiguration struct {
  KeyPath string `yaml...`
  CrtPath ... 
  CACrtPath .. 
}```

File: client/etcd/config.go [hunk 2]
```diff
@@ -26,16 +26,28 @@ import (
...
+// ClusterConfig is the config for a zoned etcd cluster.
 type ClusterConfig struct {
-	Zone      string   `yaml:"zone"`
-	Endpoints []string `yaml:"endpoints"`
-	Cert      string   `yaml:"cert",omitempty`
-	Key       string   `yaml:"key",omitempty`
-	CA        string   `yaml:"ca",omitempty`
+	Zone      string     `yaml:"zone"`
+	Endpoints []string   `yaml:"endpoints"`
+	Auth      AuthConfig `yaml:"auth"`
```
Comment (prateek) [inline client/etcd/config.go]: maybe make this type `*AuthConfig` to allow it to be nullable?

File: client/etcd/options.go [hunk 3]
```diff
@@ -21,56 +21,80 @@
...
-	SetServiceDiscoveryConfig(cfg etcdsd.Configuration) Options
+// NewAuthOptions creates a set of Auth Options.
+func NewAuthOptions() AuthOptions {
+	return authOptions{}
+}
 
-	Clusters() []Cluster
-	SetClusters(clusters []Cluster) Options
-	ClusterForZone(z string) (Cluster, bool)
+type authOptions struct {
```
Comment (prateek) [inline client/etcd/options.go]: nit: same rename suggestions as above

File: client/etcd/config.go (hunk 2 above)
Comment (cw9) [reply client/etcd/config.go]: The default struct can create a valid default option and saves a nil check, which is easier to use. Gonna keep it like this.

File: client/etcd/config.go (hunk 1 above)
Comment (cw9) [reply client/etcd/config.go]: sg

File: client/etcd/options.go (hunk 3 above)
Comment (cw9) [reply client/etcd/options.go]: sg

File: client/etcd/options.go [hunk 4]
```diff
@@ -68,30 +68,29 @@ func (o tlsOptions) SetCACrtPath(ca string) TLSOptions {
 }
 
 func (o tlsOptions) Config() (*tls.Config, error) {
-	var tlscfg *tls.Config
 	if o.cert != "" {
```
Comment (prateek) [inline client/etcd/options.go]: minor nit: could you go the other route - 

```
if o.cert == "" {
  return nil, nil 
}
// rest of the code
```

Comment (cw9) [reply client/etcd/options.go]: true, more concise that way

File: client/etcd/config.go [hunk 5]
```diff
@@ -26,16 +26,32 @@ import (
...
 
-// Configuration is for config service client
+// TLSConfig is the config for TLS.
+type TLSConfig struct {
+	CrtPath   string `yaml:"crtPath"`
+	CACrtPath string `yaml:"caCrtPath"`
+	KeyPath   string `yaml:"keyPath"`
+}
+
+func newTLSOptions(c *TLSConfig) TLSOptions {
```
Comment (xichen2020) [inline client/etcd/config.go]: nit: Convention is to do
```
func (c *TLSConfig) NewTLSOptions() {
...
}
```
and do the nil check in `NewOptions`.

File: client/etcd/types.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package etcd
+
+import (
+	"crypto/tls"
+
+	etcdsd "github.com/m3db/m3cluster/services/client/etcd"
```
Comment (prateek) [inline client/etcd/types.go]: nit: import ordering

**Pull request merged**